contracts/discovery/GNS.sol

```diff
@@ -289,40 +301,142 @@ contract GNS is GNSV2Storage, GraphUpgradeable, IGNS, Multicall {
+            if (_versionExists(subgraphData.versions[VersionType.New])) {
+                tokens += curation.burn(
+                    subgraphData.versions[VersionType.New].subgraphDeploymentID,
+                    subgraphData.versions[VersionType.New].vSignal,
+                    0
+                );
+            }
+
             address subgraphOwner = msg.sender;
+
+            // Divide tokens in half
             uint256 tokensWithTax = _chargeOwnerTax(
                 tokens,
                 subgraphOwner,
                 curation.curationTaxPercentage()
-            );
+            ).div(2);
 
-            // Update pool: constant nSignal, vSignal can change (w/no slippage protection)
-            subgraphData.subgraphDeploymentID = _subgraphDeploymentID;
-            (subgraphData.vSignal, ) = curation.mint(
+            // Mint half to old version
+            (uint256 _vSignalOld, ) = curation.mint(
                 subgraphData.subgraphDeploymentID,
                 tokensWithTax,
                 0
             );
+
+            // Mint half to new version
+            (uint256 _vSignalNew, ) = curation.mint(_subgraphDeploymentID, tokensWithTax, 0);
+
+            uint256 vSignalTotal = _vSignalOld + _vSignalNew;
+
+            // Update vSignal for versions
+            subgraphData.versions[VersionType.Current].vSignal = _vSignalOld;
+            subgraphData.versions[VersionType.New].vSignal = _vSignalNew;
+
+            // Update subgraphData Signal
+            subgraphData.vSignal = vSignalTotal;
+
             emit SubgraphUpgraded(
                 _subgraphID,
-                subgraphData.vSignal,
-                tokensWithTax,
+                vSignalTotal,
+                tokensWithTax.mul(2),
                 _subgraphDeploymentID
             );
         }
 
+        // Update versions
+        subgraphData.versions[VersionType.Current].subgraphDeploymentID = subgraphData
+            .subgraphDeploymentID;
+        subgraphData.versions[VersionType.New].subgraphDeploymentID = _subgraphDeploymentID;
+
+        // Update deployment ID
+        subgraphData.subgraphDeploymentID = _subgraphDeploymentID;
+
         emit SubgraphVersionUpdated(_subgraphID, _subgraphDeploymentID, _versionMetadata);
     }
 
+    /**
+     * @dev Finalize the upgrade process between two versions belonging to a subgraph
+     * @param _subgraphID Subgraph ID
+     */
+    function finalizeSubgraphUpgrade(uint256 _subgraphID)
+        external
+        override
+        notPaused
+        onlySubgraphAuth(_subgraphID)
+    {
+        // Subgraph check
+        SubgraphData storage subgraphData = _getSubgraphOrRevert(_subgraphID);
+
+        // Check new version exists
+        require(
+            subgraphData.versions[VersionType.New].vSignal != 0,
+            "No versions to finish upgrade"
+        );
+
+        ICuration curation = curation();
+
+        uint256 tokens;
+
+        // Burn signal
+        tokens = curation.burn(
```

Comment:
In this case, can we just burn the shares for the `current` version to save some gas?